File: backend/challenge-resolution/src/main/kotlin/core/domain/order/Order.kt
package main.kotlin.core.domain.order

import main.kotlin.core.domain.address.Address
import main.kotlin.core.domain.customer.Customer
import main.kotlin.core.domain.payment.Payment
import main.kotlin.core.domain.payment.PaymentMethod
import main.kotlin.core.domain.product.Product
import main.kotlin.core.domain.shipping.Shipment
import main.kotlin.core.exception.OrderAlreadyPayException
import main.kotlin.core.exception.OrderEmptyCanBePaidException
import main.kotlin.core.exception.ProductAlreadyAddedException
import java.util.*

class Order(val customer: Customer, val address: Address) {
    private val items = mutableListOf<OrderItem>()
    var closedAt: Date? = null
        private set
    var payment: Payment? = null
        private set
    val totalAmount
        get() = items.sumByDouble { it.total }

    fun addProduct(product: Product, quantity: Int) {
        val productAlreadyAdded = items.any { it.product == product }
        if (productAlreadyAdded)
            throw ProductAlreadyAddedException()

        items.add(OrderItem(product, quantity))
    }

    fun pay(method: PaymentMethod) {
        if (payment != null)
            throw OrderAlreadyPayException()

        if (items.count() == 0)
            throw OrderEmptyCanBePaidException()

        payment = Payment(this, method)

        close()
    }

    private fun close() {
        closedAt = Date()
    }

    fun shipItems() {
        items.forEach{ Shipment.shipProduct(it.product, customer) }
    }
}

File: backend/brbucci_challenge/order.rb
require 'pry'
require './order_item'
require './membership'
require './payment'
require './credit_card'
require './shipping'

class Order
  attr_reader :customer, :items, :payment, :address, :closed_at, :number, :shipping

  def initialize(customer, overrides = {})
    @number = 5.times.map { rand(1..9) }.join.to_i
    @customer = customer
    @items = []
    @order_item_class = overrides.fetch(:item_class) { OrderItem }
    @address = overrides.fetch(:address) { Address.new(zipcode: '45678-979') }
    return nil unless overrides.any?
    add_product(overrides[:product])
    order_payment = Payment.new(order: self, payment_method: CreditCard.fetch_by_hashed('43567890-987654367'))
    order_payment.pay
    membership = Membership.new(self) if order_payment.order.items.first.product.type.include?('membership')
    options = { member: membership.member } if membership
    process_shipping(options) if order_payment.paid?
  end

  def add_product(product)
    @items << @order_item_class.new(self, product)
  end

  def total_amount
    @items.map(&:total).inject(:+)
  end

  def process_shipping(options = {})
    @shipping = Shipping.new(self, options)
  end

  def close(closed_at = Time.now)
    @closed_at = closed_at
  end
end
